File: desktop/windows/VoiceBridgeAI/VoiceBridgeAI/Session/PcmSilenceMonitor.cs
namespace VoiceBridgeAI.Session;

/// <summary>Detect sustained silence in PCM chunks (matches server RMS VAD threshold).</summary>
public struct PcmSilenceMonitor
{
    public float RmsThreshold { get; set; }
    public TimeSpan ClearAfter { get; set; }

    private DateTime? _silentSince;

    public PcmSilenceMonitor()
    {
        RmsThreshold = 0.012f;
        ClearAfter = TimeSpan.FromSeconds(2.5);
    }

    public void Reset()
    {
        _silentSince = null;
    }

    /// <returns>True once silence has exceeded ClearAfter.</returns>
    public bool Feed(ReadOnlySpan<byte> pcm)
    {
        var rms = ComputeRms(pcm);
        if (rms >= RmsThreshold)
        {
            _silentSince = null;
            return false;
        }

        var now = DateTime.UtcNow;
        _silentSince ??= now;
        return now - _silentSince.Value >= ClearAfter;
    }

    private static float ComputeRms(ReadOnlySpan<byte> pcm)
    {
        if (pcm.Length < 2)
        {
            return 0;
        }

        var count = pcm.Length / 2;
        var sum = 0f;
        for (var i = 0; i < count; i++)
        {
            var sample = BitConverter.ToInt16(pcm.Slice(i * 2, 2)) / 32768f;
            sum += sample * sample;
        }

        return MathF.Sqrt(sum / Math.Max(count, 1));
    }
}

File: desktop/macos/Sources/VoiceBridgeAI/Session/PcmSilenceMonitor.swift
import Foundation

/// Detects sustained silence in PCM chunks (matches server RMS VAD threshold).
struct PcmSilenceMonitor {
    var rmsThreshold: Float = 0.012
    var clearAfterSeconds: TimeInterval = 2.5

    private var silentSince: Date?

    mutating func reset() {
        silentSince = nil
    }

    /// Returns true once silence has exceeded `clearAfterSeconds`.
    mutating func feed(pcm: Data) -> Bool {
        let rms = Self.rms(pcm)
        if rms >= rmsThreshold {
            silentSince = nil
            return false
        }
        let now = Date()
        if silentSince == nil {
            silentSince = now
        }
        guard let since = silentSince else { return false }
        return now.timeIntervalSince(since) >= clearAfterSeconds
    }

    private static func rms(_ pcm: Data) -> Float {
        guard pcm.count >= 2 else { return 0 }
        let count = pcm.count / 2
        var sum: Float = 0
        pcm.withUnsafeBytes { raw in
            let samples = raw.bindMemory(to: Int16.self)
            for i in 0..<count {
                let s = Float(samples[i]) / 32768
                sum += s * s
            }
        }
        return sqrt(sum / Float(max(count, 1)))
    }
}
